godot-core/src/builtin/basis.rs

```diff
@@ -0,0 +1,980 @@
+            "Fail due to X {} with {deg_original_euler} using {rot_order:?}",
+            res.col_x()
+        );
+        assert!(
+            (res.col_y() - Vector3::new(0.0, 1.0, 0.0)).length() <= 0.1,
+            "Fail due to Y {} with {deg_original_euler} using {rot_order:?}",
+            res.col_y()
+        );
+        assert!(
+            (res.col_z() - Vector3::new(0.0, 0.0, 1.0)).length() <= 0.1,
+            "Fail due to Z {} with {deg_original_euler} using {rot_order:?}",
+            res.col_z()
+        );
+
+        // Double check `to_rotation` decomposing with XYZ rotation order.
+        let euler_xyz_from_rotation: Vector3 = to_rotation.euler_angles(EulerOrder::XYZ);
+        let rotation_from_xyz_computed_euler: Basis = Basis::from_euler(
+            EulerOrder::XYZ,
+            euler_xyz_from_rotation.x,
+            euler_xyz_from_rotation.y,
+            euler_xyz_from_rotation.z,
+        );
+
+        let res = to_rotation.inverse() * rotation_from_xyz_computed_euler;
+
+        assert!(
+            (res.col_x() - Vector3::new(1.0, 0.0, 0.0)).length() <= 0.1,
+            "Double check with XYZ rot order failed, due to X {} with {deg_original_euler} using {rot_order:?}",
+            res.col_x(),
+        );
+        assert!(
+            (res.col_y() - Vector3::new(0.0, 1.0, 0.0)).length() <= 0.1,
+            "Double check with XYZ rot order failed, due to Y {} with {deg_original_euler} using {rot_order:?}",
+            res.col_y(),
+        );
+        assert!(
+            (res.col_z() - Vector3::new(0.0, 0.0, 1.0)).length() <= 0.1,
+            "Double check with XYZ rot order failed, due to Z {} with {deg_original_euler} using {rot_order:?}",
+            res.col_z(),
+        );
+    }
+
+    #[test]
+    fn consts_behavior_correct() {
+        let v = Vector3::new(1.0, 2.0, 3.0);
+
+        assert!(
+            (Basis::IDENTITY * v).is_equal_approx(v),
+            "got {got}, expected {v}",
+            got = Basis::IDENTITY * v
+        );
+        let u = Vector3::new(-v.x, v.y, v.z);
+        assert!(
+            (Basis::FLIP_X * v).is_equal_approx(u),
+            "got {got}, expected {u}",
+            got = Basis::FLIP_X * v
+        );
+        let u = Vector3::new(v.x, -v.y, v.z);
+        assert!(
+            (Basis::FLIP_Y * v).is_equal_approx(u),
+            "got {got}, expected {u}",
+            got = Basis::FLIP_X * v
+        );
+        let u = Vector3::new(v.x, v.y, -v.z);
+        assert!(
+            (Basis::FLIP_Z * v).is_equal_approx(u),
+            "got {got}, expected {u}",
+            got = Basis::FLIP_X * v
+        );
+    }
+
+    #[test]
+    fn basic_rotation_correct() {
+        assert!(
+            (Basis::from_axis_angle(Vector3::FORWARD, 0.0) * Vector3::RIGHT)
+                .is_equal_approx(Vector3::RIGHT),
+            "got {got}, expected {expected}",
+            got = Basis::from_axis_angle(Vector3::FORWARD, 0.0) * Vector3::RIGHT,
```

Comment:
Since you're repeating longer expressions, it makes sense to extract them. This would also have avoided the mistakes with `FLIP_X` above:
```rs
let actual = Basis::from_axis_angle(Vector3::FORWARD, 0.0) * Vector3::RIGHT;
let expected = Vector3::RIGHT;
assert!(actual, expected, "... what is being tested ...");
```